src/memory/cgroup_v2.rs

```diff
@@ -0,0 +1,147 @@
+use std::io::BufRead;
+use std::path::Path;
+
+use super::Memory;
+use crate::{bytes_to_kilo_bytes, file_to_buf_reader, parse_u64, read_file_value_as_u64};
+use crate::{path_to_string, ProbeError, Result};
+#[cfg(target_os = "linux")]
+pub fn read_and_parse_v2_sys_memory(path: &Path) -> Result<Memory> {
+    let mut memory = Memory {
+        total: None,
+        free: None,
+        used: 0,
+        buffers: None,
+        cached: None,
+        shmem: None,
+        swap_total: None,
+        swap_free: None,
+        swap_used: None,
+    };
+
+    memory.total = read_file_value_as_u64(&path.join("memory.max"))
+        .ok()
+        .map(bytes_to_kilo_bytes);
+
+    memory.used = bytes_to_kilo_bytes(read_file_value_as_u64(&path.join("memory.current"))?);
+
+    let reader = file_to_buf_reader(&path.join("memory.stat"))?;
+    for line_result in reader.lines() {
+        let line = line_result.map_err(|e| ProbeError::IO(e, path_to_string(path)))?;
+        let segments: Vec<&str> = line.split_whitespace().collect();
+        let value = parse_u64(&segments[1])?;
+
+        if segments[0] == "shmem" {
+            memory.shmem = Some(bytes_to_kilo_bytes(value));
+            break;
+        };
+    }
+
+    memory.free = memory.total.map(|total| total - memory.used);
+
+    memory.swap_total = match read_file_value_as_u64(&path.join("memory.swap.max")) {
+        Ok(value) => memory
+            .total
+            .map(|total| bytes_to_kilo_bytes(value).saturating_sub(total)),
```

Comment:
I don't know if this is necessary for v2 groups. If I run this command, the reported memory swap max is 100mb:

```
$ docker run --memory 200mb --memory-swap 300mb --rm -it ubuntu:22.04 cat /sys/fs/cgroup/memory.swap.max
104857600
```

In the docker command you need to specify your base memory max [+ the extra space for swap](https://docs.docker.com/config/containers/resource_constraints/#--memory-swap-details), but the reported value is `--memory 200mb` - `--memory 300mb` = 100mb swap max

Reply:
Thank you for noticing this! This is so confusing. 😅